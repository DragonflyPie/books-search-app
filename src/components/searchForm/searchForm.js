import React from "react";
import { BiSearchAlt } from "react-icons/bi";
import useSearchForm from "./useSearchForm";

const SearchForm = () => {
  const classNames = require("classnames");

  const {
    submitSearchForm,
    handleCategoryChange,
    handleOrderByChange,
    handleQueryChange,
    categoryOptions,
    orderBy,
    category,
    query,
    wrongQuery,
  } = useSearchForm();

  let inputClass = classNames({
    "search-form__input": true,
    "search-form__input--wrong": wrongQuery,
  });

  return (
    <form
      onSubmit={submitSearchForm}
      className="search-form"
      name="search-form"
    >
      <div className="search-form__input-group">
        <input
          className={inputClass}
          type="text"
          placeholder={wrongQuery ? "" : "Search..."}
          value={query}
          onChange={handleQueryChange}
        />
        <button className="search-form__btn" type="submit">
          <BiSearchAlt />
        </button>
      </div>
      {wrongQuery ? (
        <div className="search-form__error" data-testid="error">
          Missing query
        </div>
      ) : (
        ""
      )}
      <div className="search-form__bar">
        <div className="search-form__select-group">
          <label htmlFor="category">Category</label>
          <select
            className="search-form__select"
            name="category"
            data-testid="category"
            value={category}
            onChange={handleCategoryChange}
          >
            <option value="all">All</option>
            {categoryOptions}
          </select>
        </div>
        <div className="search-form__select-group">
          <label htmlFor="orderBy">Order By</label>
          <select
            className="search-form__select"
            name="orderBy"
            data-testid="orderBy"
            id="orderBy"
            value={orderBy}
            onChange={handleOrderByChange}
          >
            <option value="relevance">Relevance</option>
            <option value="newest">Newest</option>
          </select>
        </div>
      </div>
    </form>
  );
};

export default SearchForm;
