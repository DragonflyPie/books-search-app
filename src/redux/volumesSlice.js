import {
  createSlice,
  createAsyncThunk,
  createEntityAdapter,
} from "@reduxjs/toolkit";
import key from "../apiKey";

const BASE_QUERY = "https://www.googleapis.com/books/v1/volumes";

const FIELDS =
  "totalItems,items(id,volumeInfo(description,publishedDate,imageLinks,authors,categories,title))";

const createSearchQuery = (searchParams) => {
  return `${BASE_QUERY}?q=${removeSpaces(searchParams.query)}${getCategory(
    searchParams.category
  )}&orderBy=${searchParams.orderBy}&startIndex=${getPage(
    searchParams.page
  )}&maxResults=10&fields=${FIELDS}&key=${key}`;
};

const removeSpaces = (query) => {
  return query.replace(/\s/g, "+");
};

const getCategory = (category) => {
  return category === "all" ? "" : `+subject:${category}`;
};

const getPage = (page) => {
  return page * 10 - 9;
};
const volumesAdapter = createEntityAdapter({
  selectId: (volume) => volume.id,
});

const searchInitialState = volumesAdapter.getInitialState({
  status: "idle",
  error: null,
  totalItems: null,
});

export const fetchVolumes = createAsyncThunk(
  "search/volumes",
  async (_, { getState, rejectWithValue }) => {
    const searchParams = await getState().search;
    const searchQuery = await createSearchQuery(searchParams);
    const response = await fetch(searchQuery);
    const data = await response.json();
    if (response.ok === false) {
      return rejectWithValue(data.error.message);
    } else if (data.totalItems === 0) {
      return rejectWithValue("Nothing was found");
    }
    return data;
  }
);

export const fetchVolumeById = createAsyncThunk(
  "search/volume",
  async (volumeId, { rejectWithValue }) => {
    const response = await fetch(`${BASE_QUERY}/${volumeId}?key=${key}`);
    const data = await response.json();
    if (response.ok === false) {
      return rejectWithValue(data.error.message);
    }
    return data;
  }
);
const volumesSlice = createSlice({
  name: "volumes",
  initialState: searchInitialState,
  reducers: {
    resetVolumesState: () => {
      return searchInitialState;
    },
  },
  extraReducers(builder) {
    builder.addCase(fetchVolumes.pending, (state, action) => {
      state.status = "loading";
    });
    builder.addCase(fetchVolumes.rejected, (state, action) => {
      state.error = action.payload;
      state.status = "failed";
    });
    builder.addCase(fetchVolumes.fulfilled, (state, action) => {
      volumesAdapter.upsertMany(state, action.payload.items);
      state.totalItems = action.payload.totalItems;
      state.status = "succeeded";
    });
    builder.addCase(fetchVolumeById.pending, (state, action) => {
      state.status = "loading";
    });
    builder.addCase(fetchVolumeById.rejected, (state, action) => {
      state.error = action.payload;
      state.status = "failed";
    });
    builder.addCase(fetchVolumeById.fulfilled, (state, action) => {
      volumesAdapter.upsertOne(state, action.payload);
      state.status = "succeeded";
    });
  },
});

export const {
  selectAll: selectAllVolumes,
  selectById: selectVolumeById,
  selectIds: selectVolumesIds,
} = volumesAdapter.getSelectors((state) => state.volumes);

export const { resetVolumesState } = volumesSlice.actions;

export default volumesSlice.reducer;
