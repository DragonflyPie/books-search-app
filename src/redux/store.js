import { configureStore } from "@reduxjs/toolkit";
import searchReducer from "./searchSlice";
import volumesSlice from "./volumesSlice";

export const store = configureStore({
  reducer: {
    search: searchReducer,
    volumes: volumesSlice,
  },
});
