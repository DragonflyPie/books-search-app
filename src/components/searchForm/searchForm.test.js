import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import SearchForm from "./SearchForm";
import { store } from "../../redux/store";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

describe("SearchForm", () => {
  it("Inputs render", () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <SearchForm />
        </Provider>
      </BrowserRouter>
    );

    expect(screen.getByRole("textbox")).toBeInTheDocument();
    expect(screen.getByRole("form")).toBeInTheDocument();
    expect(screen.getByTestId("category")).toBeInTheDocument();
    expect(screen.getByTestId("orderBy")).toBeInTheDocument();
  });

  it("Search validates with existing query", () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <SearchForm />
        </Provider>
      </BrowserRouter>
    );

    userEvent.type(screen.getByRole("textbox"), "javascript");
    userEvent.click(screen.getByRole("button"));

    expect(screen.queryByTestId("error")).toBeNull();
  });

  it("Search not validates when query is missing", () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <SearchForm />
        </Provider>
      </BrowserRouter>
    );

    userEvent.click(screen.getByRole("button"));

    expect(screen.getByTestId("error")).toBeInTheDocument();
  });

  it("Matches snapshot", () => {
    const form = render(
      <BrowserRouter>
        <Provider store={store}>
          <SearchForm />
        </Provider>
      </BrowserRouter>
    );
    expect(form).toMatchSnapshot();
  });
});
