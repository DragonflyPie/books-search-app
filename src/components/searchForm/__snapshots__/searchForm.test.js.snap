// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm Matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="search-form"
        name="search-form"
      >
        <div
          class="search-form__input-group"
        >
          <input
            class="search-form__input"
            placeholder="Search..."
            type="text"
            value=""
          />
          <button
            class="search-form__btn"
            type="submit"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
              />
              <path
                d="M11.412 8.586c.379.38.588.882.588 1.414h2a3.977 3.977 0 0 0-1.174-2.828c-1.514-1.512-4.139-1.512-5.652 0l1.412 1.416c.76-.758 2.07-.756 2.826-.002z"
              />
            </svg>
          </button>
        </div>
        
        <div
          class="search-form__bar"
        >
          <div
            class="search-form__select-group"
          >
            <label
              for="category"
            >
              Category
            </label>
            <select
              class="search-form__select"
              data-testid="category"
              name="category"
            >
              <option
                value="all"
              >
                All
              </option>
              <option
                value="Computers"
              >
                Computers
              </option>
              <option
                value="Art"
              >
                Art
              </option>
              <option
                value="Biography"
              >
                Biography
              </option>
              <option
                value="History"
              >
                History
              </option>
              <option
                value="Science"
              >
                Science
              </option>
            </select>
          </div>
          <div
            class="search-form__select-group"
          >
            <label
              for="orderBy"
            >
              Order By
            </label>
            <select
              class="search-form__select"
              data-testid="orderBy"
              id="orderBy"
              name="orderBy"
            >
              <option
                value="relevance"
              >
                Relevance
              </option>
              <option
                value="newest"
              >
                Newest
              </option>
            </select>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="search-form"
      name="search-form"
    >
      <div
        class="search-form__input-group"
      >
        <input
          class="search-form__input"
          placeholder="Search..."
          type="text"
          value=""
        />
        <button
          class="search-form__btn"
          type="submit"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
            />
            <path
              d="M11.412 8.586c.379.38.588.882.588 1.414h2a3.977 3.977 0 0 0-1.174-2.828c-1.514-1.512-4.139-1.512-5.652 0l1.412 1.416c.76-.758 2.07-.756 2.826-.002z"
            />
          </svg>
        </button>
      </div>
      
      <div
        class="search-form__bar"
      >
        <div
          class="search-form__select-group"
        >
          <label
            for="category"
          >
            Category
          </label>
          <select
            class="search-form__select"
            data-testid="category"
            name="category"
          >
            <option
              value="all"
            >
              All
            </option>
            <option
              value="Computers"
            >
              Computers
            </option>
            <option
              value="Art"
            >
              Art
            </option>
            <option
              value="Biography"
            >
              Biography
            </option>
            <option
              value="History"
            >
              History
            </option>
            <option
              value="Science"
            >
              Science
            </option>
          </select>
        </div>
        <div
          class="search-form__select-group"
        >
          <label
            for="orderBy"
          >
            Order By
          </label>
          <select
            class="search-form__select"
            data-testid="orderBy"
            id="orderBy"
            name="orderBy"
          >
            <option
              value="relevance"
            >
              Relevance
            </option>
            <option
              value="newest"
            >
              Newest
            </option>
          </select>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
